poptorch_experimental_addons/_impl/collectives.py

```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+import torch
+import poptorch
+
+
+def replicated_all_gather(x: torch.Tensor, replication_factor: int) -> torch.Tensor:
+    """
+    SPMD all-gather across IPUs.
+
+    Gathers and concatenates tensors occupying the same memory location across IPUs,
+    then replicates the result.
+    """
```

Comment:
Perhaps avoid the SPMD acronym - could be a bit jargon-y, e.g. "All-gather across replicas."

Could be helpful to give shape details here? E.g. something like "x -- shape (*)" and "returns -- shape (replication_factor, *)`.

Reply:
> > Perhaps avoid the SPMD acronym - could be a bit jargon-y, e.g. "All-gather across replicas."
>
> Can do, although I was afraid "replicas" is a bit IPU-jargony, and SPMD would be the more generic term.

True - but we do use `replicated` in the name, and the function only works on IPUs in any case, so might still be tempted to be more explicit about it.

---

While I think of it, what do you think of `replica_all_gather` as a name? The thing about `replicated` is that I wonder what was replicated (whereas perhaps "replica" can be used to consistently mean "program replica")? Or, if we wanted to follow GCL naming, could consider `all_gather_cross_replica`.

Reply:
`all_gather_cross_replica` sounds like a sensible approach.